_posts/2015-09-29-Asking-Users-for-Help.md

```diff
@@ -0,0 +1,155 @@
+This was a programmer's nightmare! A bug keeps getting reported but you don’t have any crash logs and you have no idea how to reproduce the problem.
+
+##The Plan
+
+Giving up was not an option, so we decided to ask those for help, who actually could reproduce the problem. Our users!
+Our plan was to create a special version of the app with extensive logging. We hoped that our users would be kind enough to install this version and by doing so, help us identify the issue.
+
+##Have theories you want to verify
+
+We had several suspicions / theories which we wanted to check:
+
+- Race conditions in our cache. 
+ - <i>This is a problem that typically occurs when you have multiple threads depending on a shared state and you failed to synchronize the different threads properly. The evil thing about this issue is, that it heavily depends on the execution order of your threads. Everything can go fine a hundred times but every once in awhile your threads randomly get processor time in a specific order which causes the bug to appear. And when you have several million users this will happen a lot even if it does not happen for you!</i>
+
+- A logic problem which we just did not see when checking the code.
+
+- Some faulty error handling while syncing with the backend.
+
+- A server instance sending us corrupt data for some reason.
+
+In general it seems to be helpful if you have at least some idea where you have vulnerable parts in your code. This makes it much easier to decide what to log.
+
+##What do you want to log?
+
+To be able to properly judge which theories if any are true and wich are faulty we decided to log the following data:
+
+- user id
+
+- exact time you log something
+- The overall uptime of the device when you log something
+ - This helps you to detect restarts of the device
+- The screens the user is navigating to
+
+- The different method calls which seem to be relevant (writing into the database, starting syncs etc.)
+ - If you  suspect race conditions in specific method calls, you should log method start and method end.
+
+- The thread names / ids on which the methods were executed
+ - This helps you identify race conditions.
+
+- The states of the user settings after a change
+
+- The data we got from the backend
+
+- The data we sent to the backend
+
+####Important things you should keep in mind when you create logs:
+- Log everything you need to identify your problem. If you forget something crucial you are in trouble! Once you send your version to the user, the only thing you can do, is create a new version. But how many times can you send a new version to a user, before he decides that he is getting very annoyed by you?
+
+- Don’t log too much. At first we logged so much information that it was hard to read what was going on. By reducing the information to the important parts it became a lot easier.
```

Comment:
"Log everything but not too much" -> clearly that's a painful problem to debug :D